enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,609 @@
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "whereabouts-conf",
+    "namespace": "default",
+    "ips": [
+      "192.168.2.13/24"
+    ],
+    "interface": "pod16367aacb67",
+    "cni-args": {
+      "persistentip.cni.cncf.io/ownerID": "14709d25-22da-4ceb-9ded-d6776f19d81d",
+      "persistentip.cni.cncf.io/ownerName": "vm-rhel8",
+      "persistentip.cni.cncf.io/ownerType": "VirtualMachine",
+      "persistentip.cni.cncf.io/ownerVersion": "kubevirt.io/v1"
+    }
+  }
+]
+```
+
+When the IPAM CNI sees these CNI arguments, it will add `ownerReferences`
+metadata to the persistent IP CR, and label it with said data, to simplify
+retrieving this allocation on sub-sequent CNI operations.
+
+#### Specifying persistent IPs for a NAD
+
+Below you can find a Network Attachment Definition requesting delegated IPAM
+from a separate IPAM plugin.
+
+```yaml
+---
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: whereabouts-conf
+spec:
+  config: '{
+      "cniVersion": "0.3.1",
+      "name": "l2-network",
+      "type": "ovn-k8s-cni-overlay",
+      "topology":"layer2",
+      "mtu": 1300,
+      "netAttachDefName": "default/whereabouts-conf",
+      "ipam": {
+        "type": "whereabouts",
+        "log_file": "/var/log/wb.log",
+        "log_level": "debug",
+        "range": "192.168.2.0/24"
+      }
+    }'
+```
+
+Thus, we would have the following three behaviors:
+- cni conf without a subnet defined **and** without IPAM key defined => only
+  static IP can be requested
+- cni conf without a subnet defined **but with** the IPAM key defined =>
+  delegate IPAM to an external plugin
+- cni conf with the subnet key defined => OVN-K provided IPAM for the provided
+  subnet
+
+**NOTE**: it results in an error if the user attempts to define **both** the
+subnet **and** the IPAM key at the same time.
+
+### Risks and Mitigations
+
+The proposal is assuming the "sticky IPs" feature to be implemented in the IPAM
+plugin - e.g. whereabouts. Performance wise, this might not be a good option
+since we would need to invoke the IPAM plugin via CNI exec based API - i.e.
+invoke the CNI IPAM binary on disk, get the result (the IP), and then use said
+IP in the entire flow. Invoking the IPAM plugin is a synchronous operation that
+would block the OVN-K plugin until it gets the response back.
+
+Given whereabouts instances actively compete for a datastore lock to update the
+IP pool allocations, need to check for the presence of overlapping IPs, and
+would - on top - have to also query the datastore for the presence of existing
+persistent allocations / persist new allocations, it is probable the time it
+takes to provision a pod to grow.
+
+If we implemented the IPAM feature in OVN-Kubernetes instead (that already uses
+an asynchronous model) we would ensure a better throughput when processing the
+CNI requests.
```

Comment:
dont we want this proposal to include only what we will do in the end, without the alternatives?
if not, then maybe the alternative should be listed in a more explicit section ?